airflow/providers/airbyte/hooks/airbyte.py

```diff
@@ -0,0 +1,92 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import time
+from typing import Optional
+
+from airflow.exceptions import AirflowException
+from airflow.providers.http.hooks.http import HttpHook
+
+
+class AirbyteJobController:
+    """Airbyte job status"""
+
+    RUNNING = "running"
+    SUCCEEDED = "succeeded"
+    CANCELLED = "canceled"
+    PENDING = "pending"
+    FAILED = "failed"
+    ERROR = "error"
+
+
+class AirbyteHook(HttpHook, AirbyteJobController):
+    """Hook for Airbyte API"""
+
+    def __init__(self, airbyte_conn_id: str) -> None:
+        super().__init__(http_conn_id=airbyte_conn_id)
+
+    def wait_for_job(self, job_id: str, wait_time: int = 3, timeout: Optional[int] = None) -> None:
```

Comment:
this is very close to a sensor... might consider making a sensor so you can free up worker slot while waiting for job to complete.

Reply:
I think we may make it configurable (async/sync), that's how many of Google operators work.

Reply:
yeah like an operator that is also a sensor? like it submits the job and immediately starts sensing it?  is there an example of this you can point me to @turbaszek?  i was gonna suggest this but you have to store job_id somewhere and i know folks ran into issues before where xcom clears itself when sensor wakes up again thus purging the job id.  i'm curious how it has been implemented with google.  maybe xcom persistence issue was resolved with introduction of xcom backends?

Reply:
>  is there an example of this you can point me to @turbaszek?

Configurable waiting:
https://github.com/apache/airflow/blob/13854c32a38787af6d8a52ab2465cb6185c0b74c/airflow/providers/google/cloud/operators/dataproc.py#L895
https://github.com/apache/airflow/blob/13854c32a38787af6d8a52ab2465cb6185c0b74c/airflow/providers/google/cloud/operators/dlp.py#L263
https://github.com/apache/airflow/blob/13854c32a38787af6d8a52ab2465cb6185c0b74c/airflow/providers/google/cloud/operators/dataflow.py#L93

Waiting in operator:
https://github.com/apache/airflow/blob/13854c32a38787af6d8a52ab2465cb6185c0b74c/airflow/providers/google/cloud/operators/datafusion.py#L104
in general search for any `.result()` in GCP that does polling
https://github.com/apache/airflow/blob/13854c32a38787af6d8a52ab2465cb6185c0b74c/airflow/providers/google/cloud/operators/bigquery.py#L2143

> i'm curious how it has been implemented with google.

I assume your question is mostly about cases like reruning tasks. So the answer is that idempotence is the key... some examples:
https://github.com/apache/airflow/blob/13854c32a38787af6d8a52ab2465cb6185c0b74c/airflow/providers/google/cloud/operators/bigquery.py#L2030-L2037

This of course requires an API that allows you do some idempotency handling and this of course is not always possible. As a rule of thumb we this can be done by:
1. Generating unique id which is deterministic using (dag_id, task_id, exec_date) and includes some hash of users input
1. Reattaching to existing jobs/operations

That of course has a lot of edge cases like "reattach only to running jobs" or let users override the id of operation.

I'm personally leaning to `op >> sensor` approach but many users want to do "atomic" operations like `create` or `submit`

Reply:
Very generous of you to share so many examples -- thank you.

> I'm personally leaning to op >> sensor approach but many users want to do "atomic" operations 

Before airflow 2.0, I was really not a fan of sensors.  The task latency bothered me and I tended to try to combine more operations into a single operator.  (and this approach still has some appeal for me ) but with 2.0 and specifically the improved scheduler, there's definitely not as much reason to do so, and having the separation becomes relatively more compelling.  Although then you have the issue of coordination (e.g. xcom values...)

Reply:
> Although then you have the issue of coordination (e.g. xcom keys...)

Exactly - imho having a `async` switch gives us best of both world. To those who care for performance (yes, not all care) they may do `op >> sensor` and those who want simpler DAGs can use polling in operator 👍

Reply:
@turbaszek and @dstandish I created the **AirbyteSensor** class. At that moment I think it is working like **DataprocOperator**. I can send a parameter `asynchronous = True`. In this way, **AirbyteOperator** returns the `job_id` right after the job creation call. That way, the user can follow the job using AirbyteSensor.
The DAG would look something: `op >> sensor`.

For cases where performance and works slot aren't a problem, it's possible to use the `asynchronous = False` (default) parameter the operator will make calls to the `wait_for_job` function waiting for some result: error, canceled, completed.

I worry a little about the async calls. Airbyte is a data integration/synchronization tool. It is strange that I want to trigger subsequent tasks before getting any results from AirbyteOperator. I agree that using the `op >> sensor` is more performative in terms of allocating Airflow resources, but I think it is good to leave these two options and create examples about them.

Example Airbyte API returns from the `/connections/sync` responsible to trigger a job.
```json
{
  "job": {
    "id": 0,
    "configType": "check_connection_source",
    "configId": "string",
    "createdAt": 0,
    "updatedAt": 0,
    "status": "pending"
  },
  "attempts": [
    {
      "attempt": {
        "id": 0,
        "status": "running",
        "createdAt": 0,
        "updatedAt": 0,
        "endedAt": 0,
        "bytesSynced": 0,
        "recordsSynced": 0
      },
      "logs": {
        "logLines": [
          "string"
        ]
      }
    }
  ]
}
```

Reply:
> I worry a little about the async calls. Airbyte is a data integration/synchronization tool. It is strange that I want to trigger subsequent tasks before getting any results from AirbyteOperator. I agree that using the `op >> sensor` is more performative in terms of allocating Airflow resources, but I think it is good to leave these two options and create examples about them.

right it's largely about about freeing up airflow resources.  but also by separating, there might be a case where you'd want to clear the sensor and not retrigger the submit.